<a href="https://colab.research.google.com/github/leeryleprechaun/MSDS_Practicum_Public/blob/main/Week_2_MasterFile.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Combining the financial information with the scope three data set
#### I will then take a look at how many of the data points are missing or
#### incorrectly matched
###### * Note, I used excel to cut up the original sheet so that it includes only 2020 data for the companies


In [ ]:
# Loading in packages and the raw date
# Loading in packages and the raw date
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import tkinter

%matplotlib inline

url=''
url='https://drive.google.com/uc?id=' + url.split('/')[-2]
df_revenue = pd.read_csv(url)
df_revenue

In [ ]:
# Loading in the raw verification data that I created previously
url='https://drive.google.com/file/d/197r_i8MJcqKLMjVjUYY4pqrgjPrNPgCb/view?usp=sharing'
url='https://drive.google.com/uc?id=' + url.split('/')[-2]
df = pd.read_csv(url)
df

In [ ]:
# A list of all the companies in the revenue data
company_list = df_revenue['Company Name'].tolist()
print(len(company_list))

8959


In [ ]:
# A list of all the organizations in the emissions list
org_list = df['Organization'].tolist()
print(len(org_list))

3306


In [ ]:
# Taking a look at what companies are in both lists
complete_df = set(company_list) & set(org_list)
print(len(complete_df))

269


#### There are unfortunaetly only 269 matches in the two lists, I am also going to take a look at the other file we received that has some of the names paired up, I will load that next and repeat the above procedure

In [ ]:
# Loading in data
url=''
url='https://drive.google.com/uc?id=' + url.split('/')[-2]
code_df = pd.read_csv(url)
code_df

In [ ]:
#Taking the source_org_id and converting it to a lsit
source_org_list = code_df['source_org_id'].tolist()
print(len(source_org_list))

3303


In [ ]:
# Comparing the new lists
complete_df2 = set(source_org_list) & set(org_list)
print(len(complete_df2))

234


In [ ]:
#Taking the source_org_id and converting it to a lsit
org_id_list = code_df['Organization_ID'].tolist()
print(len(source_org_list))

3303


In [ ]:
#what does that list look like?
print(org_id_list)

[1, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 37, 38, 39, 40, 41, 42, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 61, 62, 63, 66, 67, 68, 69, 70, 71, 72, 73, 74, 74, 75, 76, 77, 78, 80, 82, 83, 84, 85, 86, 87, 87, 88, 89, 91, 92, 93, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 109, 110, 111, 112, 113, 114, 116, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 147, 148, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 164, 165, 166, 167, 168, 169, 170, 171, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 203, 204, 205, 206, 207, 208, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 223, 224, 224, 225, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 23

In [ ]:
# I need to change these values to numbers
num_list = []
for i in org_id_list:
  a = int(i)
  num_list.append(a)

In [ ]:
#What does this list look like
print(num_list)

[1, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 37, 38, 39, 40, 41, 42, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 61, 62, 63, 66, 67, 68, 69, 70, 71, 72, 73, 74, 74, 75, 76, 77, 78, 80, 82, 83, 84, 85, 86, 87, 87, 88, 89, 91, 92, 93, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 109, 110, 111, 112, 113, 114, 116, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 147, 148, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 164, 165, 166, 167, 168, 169, 170, 171, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 203, 204, 205, 206, 207, 208, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 223, 224, 224, 225, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 23

In [ ]:
#Looking to see if I get better matching on the id numbers
org_id_list_2 = df['Account number'].tolist()
print(len(org_id_list_2))

3306


In [ ]:
#What does that list look like?
print(org_id_list_2)

[843435.0, 4919.0, 860250.0, 847157.0, 44.0, 285.0, 846142.0, 47752.0, 64083.0, 860247.0, 79.0, 1535.0, 276.0, 50108.0, 97.0, 839617.0, 862400.0, 851859.0, 282.0, 64.0, 45103.0, 32498.0, 53.0, 58.0, 135.0, 17604.0, 138.0, 56327.0, 149.0, 143.0, 21318.0, 154.0, 156.0, 860927.0, 313.0, 316.0, 169.0, 836137.0, 852209.0, 51887.0, 836997.0, 841759.0, 21325.0, 832235.0, 36842.0, 220.0, 21327.0, 21380.0, 28842.0, 47208.0, 333.0, 33597.0, 29552.0, 240.0, 40175.0, 862506.0, 854393.0, 334.0, 21330.0, 246.0, 60942.0, 248.0, 31416.0, 252.0, 57666.0, 258.0, 261.0, 863247.0, 256.0, 64076.0, 37280.0, 365.0, 17116.0, 362.0, 830189.0, 31144.0, 993.0, 22733.0, 383.0, 391.0, 838222.0, 351.0, 350.0, 834528.0, 21338.0, 600.0, 54751.0, 71203.0, 427.0, 429.0, 432.0, 435.0, 834609.0, 21770.0, 838754.0, 843844.0, 399.0, 403.0, 394.0, 454.0, 54202.0, 449.0, 35233.0, 21112.0, 457.0, 64656.0, 21306.0, 21388.0, 31563.0, 73490.0, 16615.0, 29174.0, 63523.0, 74116.0, 528.0, 832307.0, 524.0, 58753.0, 837003.0, 23612.0

In [ ]:
# I need to get rid of the decimal place to compare the lists correctly
new_list = []
for i in org_id_list_2:
  a = int(i)
  new_list.append(a)



In [ ]:
print(new_list)

[843435, 4919, 860250, 847157, 44, 285, 846142, 47752, 64083, 860247, 79, 1535, 276, 50108, 97, 839617, 862400, 851859, 282, 64, 45103, 32498, 53, 58, 135, 17604, 138, 56327, 149, 143, 21318, 154, 156, 860927, 313, 316, 169, 836137, 852209, 51887, 836997, 841759, 21325, 832235, 36842, 220, 21327, 21380, 28842, 47208, 333, 33597, 29552, 240, 40175, 862506, 854393, 334, 21330, 246, 60942, 248, 31416, 252, 57666, 258, 261, 863247, 256, 64076, 37280, 365, 17116, 362, 830189, 31144, 993, 22733, 383, 391, 838222, 351, 350, 834528, 21338, 600, 54751, 71203, 427, 429, 432, 435, 834609, 21770, 838754, 843844, 399, 403, 394, 454, 54202, 449, 35233, 21112, 457, 64656, 21306, 21388, 31563, 73490, 16615, 29174, 63523, 74116, 528, 832307, 524, 58753, 837003, 23612, 50132, 853114, 40183, 586, 593, 597, 43995, 569, 21355, 846879, 850298, 573, 7616, 631, 21365, 622, 625, 629, 47212, 638, 618, 628, 851037, 609, 31917, 849236, 668, 674, 21368, 641, 661, 862718, 660, 38267, 40188, 689, 692, 863036, 699, 7

In [ ]:
print(org_id_list)

[1, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 37, 38, 39, 40, 41, 42, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 61, 62, 63, 66, 67, 68, 69, 70, 71, 72, 73, 74, 74, 75, 76, 77, 78, 80, 82, 83, 84, 85, 86, 87, 87, 88, 89, 91, 92, 93, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 109, 110, 111, 112, 113, 114, 116, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 147, 148, 150, 151, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 164, 165, 166, 167, 168, 169, 170, 171, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 203, 204, 205, 206, 207, 208, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 223, 224, 224, 225, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 23

In [ ]:
# Comparing the new lists
list_compare = set(new_list) & set(num_list)
print(len(complete_df2))

234


###### 234 matches, so not much better considering we have 3,300 organizations to match up!

# Going to play around with the difflib — Helpers for computing deltas and try to get some more matching to occur
###### https://docs.python.org/3/library/difflib.html#

In [ ]:
# Loading in the package
import difflib as dl

In [ ]:
# Testing the package to make sure that it is loaded correctly
dl.get_close_matches('appel', ['ape', 'apple', 'peach', 'puppy'], n=3, cutoff=0.6)

['apple', 'ape']

In [ ]:
# Testing the package to make sure that it is loaded correctly
dl.get_close_matches('appel', ['ape', 'apple', 'peach', 'puppy'], n=3, cutoff=.2)

['apple', 'ape', 'puppy']

In [ ]:
# Testing the package to make sure that it is loaded correctly
dl.get_close_matches('appel', ['ape', 'apple', 'peach', 'puppy'], n=1, cutoff=0.6)

['apple']

##### Now that the package is working I am going to test it on the dataset

In [ ]:
#Checking on the scope 3 emissions data set
df.head(2)

In [ ]:
# Checkingon the revenue data frame
df_revenue.head(2)

In [ ]:
# Trying to check a Company Name against the revenue list
# First I am going to make a list of all the company names in the Revenue file
company_name_list = df_revenue['Company Name'].tolist()
print(len(company_name_list))

8959


In [ ]:
# Pulling out a single name from the organization lsit
name = df['Organization'].iloc[0]
print(name)

(ACIP) Alexandria Company for Industrial Packages


In [ ]:
# Trying to find a company from the scope three emissions list in the revnue file
print("The company name was: " + name)
print("The suggested name list includes:")
dl.get_close_matches(name, company_name_list, n=6, cutoff=0.2)



In [ ]:
# Trying another business name
name = df['Organization'].iloc[1]
print("The company name was: " + name)
print("The suggested name list includes:")
dl.get_close_matches(name, company_name_list, n=6, cutoff=0.2)

The company name was: 1&1 AG
The suggested name list includes:


['TUI AG', 'Do & Co AG', 'BKW AG', 'AMS AG', '4SC AG', 'Sika AG']

In [ ]:
# Changing some of the tunning for the name matching
name = df['Organization'].iloc[2]
print("The company name was: " + name)
print("The suggested name list includes:")
dl.get_close_matches(name, company_name_list, n=3, cutoff=0.3)

The company name was: 2BM Limited
The suggested name list includes:


['Bionomics Limited', 'Afterpay Limited', 'China Mobile Limited']

In [ ]:
# Building this into a loop
i = 1 
while i < 30:
  name = df['Organization'].iloc[i]
  print("The company name was: " + name)
  print("The suggested name list includes:")
  print(dl.get_close_matches(name, company_name_list, n=3, cutoff=0.3))
  print()
  print()
  i = i + 1

The company name was: 1&1 AG
The suggested name list includes:
['TUI AG', 'Do & Co AG', 'BKW AG']


The company name was: 2BM Limited
The suggested name list includes:
['Bionomics Limited', 'Afterpay Limited', 'China Mobile Limited']


The company name was: 37 Interactive Entertainment
The suggested name list includes:
['Inspired Entertainment Inc', 'Flutter Entertainment plc', 'Gungho Online Entertainment']


The company name was: 3i Group
The suggested name list includes:
['3i Group Plc', 'SJW Group', 'GPT Group']


The company name was: 3M Company
The suggested name list includes:
['3M Co', 'Telia Company AB', 'Comany Inc.']


The company name was: 3MD INC
The suggested name list includes:
['3M Co', 'cbdMD Inc', 'M3 Inc']


The company name was: 3NOD
The suggested name list includes:
['NOV Inc', 'NIO Inc', 'NOF Corp']


The company name was: 3Sbio
The suggested name list includes:
['SimCorp', 'Siix Corp', 'Tabio Corp']


The company name was: 4SERVICE KANTINE AS
The suggested name l

In [ ]:
# Playing with the tunning of difflib package, lets run it with just two suggestions
i = 1 
while i < 30:
  name = df['Organization'].iloc[100 + i]
  print("The company name was: " + name)
  print("The suggested name list includes:")
  print(dl.get_close_matches(name, company_name_list, n=2, cutoff=0.3))
  print()
  print()
  i = i + 1

The company name was: AKBANK T.A.Ş.
The suggested name list includes:
['AKBANK', 'YA-MAN LTD.']


The company name was: AKÇANSA ÇİMENTO SANAYİ VE TİCARET A.Ş.
The suggested name list includes:
['NEXTAGE Co., Ltd.', 'ASSA ABLOY AB']


The company name was: AKENERJİ ELEKTRİK ÜRETİM A.Ş.
The suggested name list includes:
['AEW UK REIT plc', 'KENKO Mayonnaise Co., Ltd.']


The company name was: Aker ASA
The suggested name list includes:
['Atea ASA', 'Akastor ASA']


The company name was: Aker BP ASA
The suggested name list includes:
['Akastor ASA', 'Atea ASA']


The company name was: Aker Solutions ASA
The suggested name list includes:
['Aker Solutions ASA (New)', 'Skyworks Solutions Inc']


The company name was: Alamos Gold Inc.
The suggested name list includes:
['Alamos Gold Inc (New)', 'Alamo Group, Inc.']


The company name was: ALBARAKA TÜRK KATILIM BANKASI A.Ş.
The suggested name list includes:
['DNB BANK ASA', 'TIM S.A.']


The company name was: Albertsons Companies, Inc.
The sugges

In [ ]:
# Playing with the tunning of difflib package, lets run it with just one suggestion 
i = 1 
while i < 10:
  name = df['Organization'].iloc[100 + i]
  print("The company name was: " + name)
  print("The suggested name list includes:")
  print(dl.get_close_matches(name, company_name_list, n=1, cutoff=0.3))
  print()
  print()
  i = i + 1

The company name was: AKBANK T.A.Ş.
The suggested name list includes:
['AKBANK']


The company name was: AKÇANSA ÇİMENTO SANAYİ VE TİCARET A.Ş.
The suggested name list includes:
['NEXTAGE Co., Ltd.']


The company name was: AKENERJİ ELEKTRİK ÜRETİM A.Ş.
The suggested name list includes:
['AEW UK REIT plc']


The company name was: Aker ASA
The suggested name list includes:
['Atea ASA']


The company name was: Aker BP ASA
The suggested name list includes:
['Akastor ASA']


The company name was: Aker Solutions ASA
The suggested name list includes:
['Aker Solutions ASA (New)']


The company name was: Alamos Gold Inc.
The suggested name list includes:
['Alamos Gold Inc (New)']


The company name was: ALBARAKA TÜRK KATILIM BANKASI A.Ş.
The suggested name list includes:
['DNB BANK ASA']


The company name was: Albertsons Companies, Inc.
The suggested name list includes:
['Patterson Companies Inc']




In [ ]:
# Playing with the tunning of difflib package, lets run it with a tighter cutoff
i = 1 
while i < 10:
  name = df['Organization'].iloc[100 + i]
  print("The company name was: " + name)
  print("The suggested name list includes:")
  print(dl.get_close_matches(name, company_name_list, n=1, cutoff=0.6))
  print()
  print()
  i = i + 1

The company name was: AKBANK T.A.Ş.
The suggested name list includes:
['AKBANK']


The company name was: AKÇANSA ÇİMENTO SANAYİ VE TİCARET A.Ş.
The suggested name list includes:
[]


The company name was: AKENERJİ ELEKTRİK ÜRETİM A.Ş.
The suggested name list includes:
[]


The company name was: Aker ASA
The suggested name list includes:
['Atea ASA']


The company name was: Aker BP ASA
The suggested name list includes:
['Akastor ASA']


The company name was: Aker Solutions ASA
The suggested name list includes:
['Aker Solutions ASA (New)']


The company name was: Alamos Gold Inc.
The suggested name list includes:
['Alamos Gold Inc (New)']


The company name was: ALBARAKA TÜRK KATILIM BANKASI A.Ş.
The suggested name list includes:
[]


The company name was: Albertsons Companies, Inc.
The suggested name list includes:
['Patterson Companies Inc']




### Taking a look at 10 companies there were 3 correct matches, 3 false matches, and then 4 blanks

In [ ]:
# Going to run the same 10 companies with a less strict cutoff
i = 1 
while i < 10:
  name = df['Organization'].iloc[100 + i]
  print("The company name was: " + name)
  print("The suggested name list includes:")
  print(dl.get_close_matches(name, company_name_list, n=1, cutoff=0.4))
  print()
  print()
  i = i + 1

The company name was: AKBANK T.A.Ş.
The suggested name list includes:
['AKBANK']


The company name was: AKÇANSA ÇİMENTO SANAYİ VE TİCARET A.Ş.
The suggested name list includes:
[]


The company name was: AKENERJİ ELEKTRİK ÜRETİM A.Ş.
The suggested name list includes:
['AEW UK REIT plc']


The company name was: Aker ASA
The suggested name list includes:
['Atea ASA']


The company name was: Aker BP ASA
The suggested name list includes:
['Akastor ASA']


The company name was: Aker Solutions ASA
The suggested name list includes:
['Aker Solutions ASA (New)']


The company name was: Alamos Gold Inc.
The suggested name list includes:
['Alamos Gold Inc (New)']


The company name was: ALBARAKA TÜRK KATILIM BANKASI A.Ş.
The suggested name list includes:
[]


The company name was: Albertsons Companies, Inc.
The suggested name list includes:
['Patterson Companies Inc']




#### This did not help the model make in more correct matches
#### I am thinking it may be best to exclude words like "Corporation" and "Inc" and corp, llc, etc., from the list before we try to match this way

#### It also appears that there is a way to compare two lists of names using difflib, so I am going to take a look at that as well



In [ ]:
# I already have alist of all the company names in the revenue file, now I need one for all the organization names in the emissions file
org_list = df['Organization'].tolist()
print(len(org_list))


3306


In [ ]:
# Now to compare the matches of those two lists
diff = dl.ndiff(org_list,company_name_list)

for l in diff:
    print(l)

#### Now the challenge will be matching up the names in a meaningful way between the two lists to add the revenue data in bulk, if we are not able to do this, and the numbers of matches is only Kellen's 850 ish we are going to be losing a massive amount of the data

# Double Checking Kellen's matching


In [ ]:
# This is merging the "Code data" with the emissions data set
# the "Code data is again the file that contains name 'Keys'""
merge_1 = pd.merge(df, code_df, left_on='Organization', right_on = "Name")
df_revenue_merged = pd.merge(merge_1, df_revenue, left_on='source_org_id', right_on = "Company Name")
df_revenue_merged = df_revenue_merged.drop(columns = ['Organization_ID', 'source_org_id', 'Name'])
df_revenue_merged.drop(df_revenue_merged.columns[[0]], axis = 1)

#### To me is looks like that matching works well, lets make a new column to compate the amount of emissions to the amount of revenue


In [ ]:
# Creating the new column to hold the ratio
df_master = df_revenue_merged
df_master['revenue_vs_emissions'] = (df_master['Revenue (in $USD)'] / df_master['total_scope_3_emissions'])
df_master.head(3)

In [ ]:
# Sorting the data frame based on the ratio column
worst = df_master.sort_values('revenue_vs_emissions')
worst.head(10)

In [ ]:
# Sorting the data frame based on the ratio column
best = df_master.sort_values('revenue_vs_emissions')
best.tail(10)